core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactory.java

```diff
@@ -308,6 +308,29 @@ RelDataType createDecimalQuotient(
   @SuppressWarnings("deprecation")
   FieldInfoBuilder builder();
 
+  /**
+   * Infers the return type of a decimal addition. Decimal addition involves
+   * at least one decimal operand and requires both operands to have exact
+   * numeric types.
+   *
+   * @param type1 type of the first operand
+   * @param type2 type of the second operand
+   * @return the result type for a decimal addition, or null if decimal
+   * addition should not be applied to the operands.
+   */
+  RelDataType createDecimalAddition(RelDataType type1, RelDataType type2);
```

Comment:
should it be a default method to preserve source compatibility?

Reply:
very good point. it should be. moving the implementation to the interface. thanks !